### PR TITLE
feat(host-rules): prepend scheme when missing one and containing a url path

### LINKED ISSUE #17918: Fail or fix matchHost values missing https:// but with a path
### What would you like Renovate to be able to do?

Automatically fix this type of config mistake: https://github.com/renovatebot/renovate/discussions/17875#discussioncomment-3709465

### If you have any ideas on how this should be implemented, please tell us here.

If we think we can always prepend `https://` in such cases (e.g. if it contains a `/` and doesn't start with `https://` then prefix it) then let's just do so (and update docs). Else we should raise a config error.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Prepend `https://` on `matchHost` urls which missing scheme but containing a path.

## Context
Closes #17918

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Can it be done with config migration instead?

**Comment by Gabriel-Ladzaretti** [comment]
```
 WARN: Config needs migrating
       "originalConfig": {
         ...
     --> "hostRules": [{"matchHost": "domain.com/", "token": "***********"}], 
         "repositoryCache": "enabled",                                                                                                                
         "repositoryCacheType": "local",                                                                                                              
         "dependencyDashboardFooter": "\n- [ ] <!-- manual job -->Check this box to trigger a request for Renovate to run again on this repository\n",
         "requireConfig": true,                                                                                                                       
         "includeForks": true,                                                                                                                        
         "dependencyDashboard": true,                                                                                                                 
         "repositories": ["ladzaretti/renovate-reproduction-12068"],                                                                                  
         "logFile": "/tmp/renovate-17-10-2022--13-32-02.log",                                                                                         
         "logFileLevel": "debug"                                                                                                                      
       },                                                                                                                                             
       "migratedConfig": {                                                                                                                            
         ...
     --> "hostRules": [{"matchHost": "https://domain.com/", "token": "***********"}],
         "repositoryCache": "enabled",
         "repositoryCacheType": "local",
         "dependencyDashboardFooter": "\n- [ ] <!-- manual job -->Check this box to trigger a request for Renovate to run again on this repository\n",
         "requireConfig": "required",
         "includeForks": true,
         "dependencyDashboard": true,
         "repositories": ["ladzaretti/renovate-reproduction-12068"],
         "logFile": "/tmp/renovate-17-10-2022--13-32-02.log",
         "logFileLevel": "debug"
       }

```

File: lib/config/migrations/custom/host-rules-migration.ts [hunk 1]
```diff
@@ -25,17 +33,25 @@ export class HostRulesMigration extends AbstractMigration {
...
-      return newRule;
-    });
+      newHostRules.push(newRule);
+    }
 
     this.rewrite(newHostRules);
   }
 }
+
+function massageUrl(url: string): string {
```
Comment (viceice) [inline lib/config/migrations/custom/host-rules-migration.ts:51]: shouldn't this be a private static class method? 🤔

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/config/migrations/custom/host-rules-migration.ts (hunk 1 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/migrations/custom/host-rules-migration.ts:51]: > Consider converting private methods to non-exported functions within the same file but outside of any class, and moving private properties into a separate, non-exported class.


https://google.github.io/styleguide/tsguide.html#visibility

but im not zealot about it :)

Comment (viceice) [reply lib/config/migrations/custom/host-rules-migration.ts:51]: ok. 😅

**Review by viceice — APPROVED** [review]

**Pull request merged**